File: runtime/cpp/titan-ast-runtime-lib/KeyWordAutomata.cpp
//
// Created by tian wei jun on 2024/1/9.
//

#include "KeyWordAutomata.h"

KeyWordAutomata::KeyWordAutomata()
    : emptyOrNot(KeyWordAutomata::EMPTY), rootKeyWord(nullptr),
      textTerminalMap(
          std::unordered_map<std::string *, Grammar *, TextTerminalMapHash,
                             TextTerminalMapEq>()) {}

KeyWordAutomata::KeyWordAutomata(int isEmpty, const Grammar *rootKeyWord)
    : emptyOrNot(isEmpty), rootKeyWord(rootKeyWord),
      textTerminalMap(
          std::unordered_map<std::string *, Grammar *, TextTerminalMapHash,
                             TextTerminalMapEq>()) {}

KeyWordAutomata::~KeyWordAutomata() = default;

std::vector<Token *> *
KeyWordAutomata::buildToken(std::vector<Token *> *tokens) const {
  for (auto token : *tokens) {
    if (rootKeyWord->index == token->terminal.index) {
      auto findIt = textTerminalMap.find(&token->text);
      if (findIt != textTerminalMap.end()) {
        auto terminal = findIt->second;
        token->terminal = *terminal;
      }
    }
  }
  return tokens;
}

File: src/titan/ast/grammar/token/KeyWordAutomata.java
package titan.ast.grammar.token;

import java.util.HashMap;
import titan.ast.grammar.Grammar;

/**
 * 只保留依赖的数据，真正的实现在runtime里面.
 *
 * @author tian wei jun
 */
public class KeyWordAutomata {
  public static final int EMPTY = 0;
  public static final int NOT_EMPTY = 1;

  public int emptyOrNot = EMPTY;

  public Grammar rootKeyWord = null;

  public HashMap<String, Grammar> textTerminalMap = new HashMap<>();
}
